test/src/vfs_test_helper.cc

```diff
@@ -0,0 +1,256 @@
+      TILEDB_OK);
+  return Status::Ok();
+}
+
+Status SupportedFsAzure::init(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) {
+  int is_container = 0;
+  int rc = tiledb_vfs_is_bucket(ctx, vfs, container.c_str(), &is_container);
+  REQUIRE(rc == TILEDB_OK);
+  if (!is_container) {
+    rc = tiledb_vfs_create_bucket(ctx, vfs, container.c_str());
+    REQUIRE(rc == TILEDB_OK);
+  }
+  return Status::Ok();
+}
+
+Status SupportedFsAzure::close(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) {
+  int is_container = 0;
+  int rc = tiledb_vfs_is_bucket(ctx, vfs, container.c_str(), &is_container);
+  CHECK(rc == TILEDB_OK);
+  if (is_container) {
+    CHECK(tiledb_vfs_remove_bucket(ctx, vfs, container.c_str()) == TILEDB_OK);
+  }
+  return Status::Ok();
+}
+
+std::string SupportedFsAzure::temp_dir() {
+  return temp_dir_;
+}
+
+Status SupportedFsLocal::prepare_config(
+    tiledb_config_t* config, tiledb_error_t* error) {
+  (void)config;
+  (void)error;
+  return Status::Ok();
+}
+
+Status SupportedFsLocal::init(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) {
+  (void)ctx;
+  (void)vfs;
+  return Status::Ok();
+}
+
+Status SupportedFsLocal::close(tiledb_ctx_t* ctx, tiledb_vfs_t* vfs) {
+  (void)ctx;
+  (void)vfs;
+  return Status::Ok();
+}
+
+#ifdef _WIN32
+// Windows local filesystem
+std::string SupportedFsLocal::temp_dir() {
+  return temp_dir_;
+}
+std::string SupportedFsLocal::file_prefix() {
+  return file_prefix_;
+}
+#else
+std::string SupportedFsLocal::temp_dir() {
+  return temp_dir_;
+}
+// Posix local filesystem
+std::string SupportedFsLocal::file_prefix() {
+  return file_prefix_;
+}
+#endif
+
+std::vector<std::unique_ptr<SupportedFs>> vfs_test_get_fs_vec() {
+  std::vector<std::unique_ptr<SupportedFs>> fs_vec;
+  bool supports_s3_ = false;
+  bool supports_hdfs_ = false;
+  bool supports_azure_ = false;
+  get_supported_fs(&supports_s3_, &supports_hdfs_, &supports_azure_);
+  if (supports_s3_) {
+    SupportedFsS3* s3_fs = new SupportedFsS3();
+    fs_vec.emplace_back(s3_fs);
+  }
+  if (supports_hdfs_) {
+    SupportedFsHDFS* hdfs_fs = new SupportedFsHDFS();
+    fs_vec.emplace_back(hdfs_fs);
+  }
+  if (supports_azure_) {
+    SupportedFsAzure* azure_fs = new SupportedFsAzure();
+    fs_vec.emplace_back(azure_fs);
+  }
+  SupportedFsLocal* local_fs = new SupportedFsLocal();
+  fs_vec.emplace_back(local_fs);
+
+  return fs_vec;
+}
+
+Status vfs_test_init(
+    const std::vector<std::unique_ptr<SupportedFs>>& fs_vec,
+    tiledb_ctx_t** ctx,
+    tiledb_vfs_t** vfs,
+    tiledb_config_t* config) {
+  tiledb_error_t* error = nullptr;
+  REQUIRE(tiledb_config_alloc(&config, &error) == TILEDB_OK);
+  REQUIRE(error == nullptr);
+  for (auto& supported_fs : fs_vec) {
+    REQUIRE(supported_fs->prepare_config(config, error).ok());
+  }
+  REQUIRE(tiledb_ctx_alloc(config, ctx) == TILEDB_OK);
+  REQUIRE(error == nullptr);
+  REQUIRE(tiledb_vfs_alloc(*ctx, config, vfs) == TILEDB_OK);
+  tiledb_config_free(&config);
```

Comment:
Don't free this if it was non-null when entering the function. If you've refactored to use `config_tmp`, just call this on `config_tmp` if the original `config` is was null.
```
if (config == nullptr)
  tiledb_config_free(&config);
```